### PR TITLE
Restore chain if reorg block fails validation

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- During chain reorganisation the original chain is restored if one of the orphan blocks in the re-org chain doesn't pass validation. 
- The Orphan block that didn't pass validation is also removed from the block orphan pool.

## Motivation and Context
This will ensure the current valid chain is not removed when an invalid fork chain is encountered.

## How Has This Been Tested?
Covered by current tests.

## Types of changes

* [x] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 1]
```diff
@@ -1283,14 +1270,43 @@ fn reorganize_chain<T: BlockchainBackend>(
...
-        let orphan_hash = block.hash();
-        txn.delete(DbKey::OrphanBlock(orphan_hash));
-        validate_and_store_new_block(metadata, db, block_validator, block)?;
+    // Validate and add chain block
+    let mut validation_result: Result<(), ValidationError> = Ok(());
+    let mut orphan_hashes = Vec::<BlockHash>::with_capacity(chain.len());
+    for block in chain {
+        orphan_hashes.push(block.hash());
+        validation_result = block_validator.validate(&block, db, metadata);
+        if validation_result.is_err() {
```
Comment (mikethetike) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: Maybe log the error here?

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 2]
```diff
@@ -1283,14 +1270,43 @@ fn reorganize_chain<T: BlockchainBackend>(
...
+                for orphan_hash in orphan_hashes {
+                    txn.delete(DbKey::OrphanBlock(orphan_hash));
+                }
+                commit(db, txn)?;
+            }
+            Ok(removed_blocks)
+        },
+        Err(e) => {
+            // Restore previous chain
+            let _ = rewind_to_height(metadata, db, height)?;
```
Comment (mikethetike) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: Maybe log which blocks were unwounded here

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 1 above)
Comment (neonknight64) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: Change applied.

File: base_layer/core/src/chain_storage/blockchain_database.rs (hunk 2 above)
Comment (neonknight64) [reply base_layer/core/src/chain_storage/blockchain_database.rs]: Change applied.

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 3]
```diff
@@ -1283,14 +1270,57 @@ fn reorganize_chain<T: BlockchainBackend>(
...
+                let mut txn = DbTransaction::new();
+                for orphan_hash in orphan_hashes {
+                    txn.delete(DbKey::OrphanBlock(orphan_hash));
+                }
+                commit(db, txn)?;
+            }
+            Ok(removed_blocks)
+        },
+        Err(e) => {
+            trace!(target: LOG_TARGET, "Restoring previous chain after failed reorg.",);
```
Comment (CjS77) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: info at least

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 4]
```diff
@@ -1283,14 +1270,57 @@ fn reorganize_chain<T: BlockchainBackend>(
...
+            );
+            remove_orphan(db, block.hash())?;
+            break;
+        }
+        store_new_block(db, block)?;
+    }
+
+    match validation_result {
+        Ok(_) => {
+            trace!(target: LOG_TARGET, "Removing reorged orphan blocks.",);
```
Comment (CjS77) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: debug

File: base_layer/core/src/chain_storage/blockchain_database.rs [hunk 5]
```diff
@@ -1283,14 +1270,57 @@ fn reorganize_chain<T: BlockchainBackend>(
...
-        validate_and_store_new_block(metadata, db, block_validator, block)?;
+    trace!(target: LOG_TARGET, "Validate and add chain blocks.",);
+    let mut validation_result: Result<(), ValidationError> = Ok(());
+    let mut orphan_hashes = Vec::<BlockHash>::with_capacity(chain.len());
+    for block in chain {
+        let block_hash = block.hash();
+        orphan_hashes.push(block_hash.clone());
+        validation_result = block_validator.validate(&block, db, metadata);
+        if validation_result.is_err() {
+            debug!(
```
Comment (CjS77) [inline base_layer/core/src/chain_storage/blockchain_database.rs]: possibly warn

**Pull request merged**